Add bootstrap and deployment scripts

marrobi opened the pull request:
# PR for issue

## What is being addressed

Add scripts to bootstrap TF environment, build and push docker images, and deploy the TRE.

## How is this addressed

- Add TF bootstrap scripts and templates to create TF state storage and ACR
- Add scripts to build and push the docker images
- Add deployment scripts to core TF templates
- Remove GitHub packages integration
- Set linter to run locally as we have already checkout out the files

deniscep commented on docs/quickstart.md:
```diff
@@ -45,29 +32,74 @@ az account list
+
+- From bash run `./deploy.sh`
+
+This Terraform creates & configures the following:
+
+- Resource Group (also in bootstrap).
+- Storage Account for holding Terraform state (also in bootstrap).
+- Azure Container Registry.
+
+### Build and push docker images
+
+To build and push the docker images run:
+
+- From bash run `./build_and_publish_docker_images.sh.sh`
+
+This Terraform creates & configures the following:
+
+- Resource Group (also in bootstrap).
+- Storage Account for holding Terraform state (also in bootstrap).
+- Azure Container Registry.
```
Looking at the code, this documentation doesn't match what the script does (build_and_publish.. doesn't create rg, etc.)?

marrobi replied:
Sorry, more haste...

marrobi replied:
Fixed

deniscep commented on templates/core/terraform/deploy.sh:
```diff
@@ -0,0 +1,73 @@
+if [ ! -f $PATH_STATE_ENV ]; then
+  echo -e "\e[31m»»» 💥 Unable to find bootstrap .env file, please create file and try again!"
+  exit
+else
+  echo -e "\n\e[34m»»» 🧩 \e[96mLoading bootstrap environmental variables\e[0m..."
+  export $(egrep -v '^#' $PATH_STATE_ENV | xargs)
+fi
+
+# Load env vars from .env file
+if [ ! -f ".env" ]; then
+  echo -e "\e[31m»»» 💥 Unable to find core .env file, please create file and try again!"
+  exit
+else
+  echo -e "\n\e[34m»»» 🧩 \e[96mLoading core environmental variables\e[0m..."
+  export $(egrep -v '^#' ".env" | xargs)
+fi
+
+az > /dev/null 2>&1
+if [ $? -ne 0 ]; then
+  echo -e "\e[31m»»» ⚠️ Azure CLI is not installed! 😥 Please go to http://aka.ms/cli to set it up"
+  exit
+fi
+
+terraform version > /dev/null 2>&1
+if [ $? -ne 0 ]; then
+  echo -e "\e[31m»»» ⚠️ Terraform is not installed! 😥 Please go to https://www.terraform.io/downloads.html to set it up"
+  exit
+fi
+
+export SUB_NAME=$(az account show --query name -o tsv)
+if [[ -z $SUB_NAME ]]; then
+  echo -e "\n\e[31m»»» ⚠️ You are not logged in to Azure!"
+  exit
+fi
+export TENANT_ID=$(az account show --query tenantId -o tsv)
+
+echo -e "\e[34m»»» 🔨 \e[96mAzure details from logged on user \e[0m"
+echo -e "\e[34m»»»   • \e[96mSubscription: \e[33m$SUB_NAME\e[0m"
+echo -e "\e[34m»»»   • \e[96mTenant:       \e[33m$TENANT_ID\e[0m\n"
+
```
same, copy-paste

deniscep commented on templates/core/terraform/destroy.sh:
```diff
@@ -0,0 +1,73 @@
+if [ ! -f $PATH_STATE_ENV ]; then
+  echo -e "\e[31m»»» 💥 Unable to find bootstrap .env file, please create file and try again!"
+  exit
+else
+  echo -e "\n\e[34m»»» 🧩 \e[96mLoading bootstrap environmental variables\e[0m..."
+  export $(egrep -v '^#' $PATH_STATE_ENV | xargs)
+fi
+
+# Load env vars from .env file
+if [ ! -f ".env" ]; then
+  echo -e "\e[31m»»» 💥 Unable to find core .env file, please create file and try again!"
+  exit
+else
+  echo -e "\n\e[34m»»» 🧩 \e[96mLoading core environmental variables\e[0m..."
+  export $(egrep -v '^#' ".env" | xargs)
+fi
+
+az > /dev/null 2>&1
+if [ $? -ne 0 ]; then
+  echo -e "\e[31m»»» ⚠️ Azure CLI is not installed! 😥 Please go to http://aka.ms/cli to set it up"
+  exit
+fi
+
+terraform version > /dev/null 2>&1
+if [ $? -ne 0 ]; then
+  echo -e "\e[31m»»» ⚠️ Terraform is not installed! 😥 Please go to https://www.terraform.io/downloads.html to set it up"
+  exit
+fi
+
+export SUB_NAME=$(az account show --query name -o tsv)
+if [[ -z $SUB_NAME ]]; then
+  echo -e "\n\e[31m»»» ⚠️ You are not logged in to Azure!"
+  exit
+fi
+export TENANT_ID=$(az account show --query tenantId -o tsv)
+
+echo -e "\e[34m»»» 🔨 \e[96mAzure details from logged on user \e[0m"
+echo -e "\e[34m»»»   • \e[96mSubscription: \e[33m$SUB_NAME\e[0m"
+echo -e "\e[34m»»»   • \e[96mTenant:       \e[33m$TENANT_ID\e[0m\n"
+
```
copy-paste

deniscep reviewed (changes requested):
Should we consider modularizing sh scripts, so we have less code? Following instructions the bootstrap.sh fails.

marrobi commented on bootstrap/terraform/.env.sample:
```diff
@@ -0,0 +1,7 @@
+TF_VAR_state_container=tfstate
+TF_VAR_location=__CHANGE_ME__
+TF_VAR_resource_name_prefix=__CHANGE_ME__
+`TF_VAR_image_tag=__CHANGE_ME__
```
Weird, didn't show in my editor. Fixed.

marrobi commented on bootstrap/terraform/bootstrap.sh:
```diff
@@ -0,0 +1,82 @@
+#!/bin/bash
```
Fixed, are you ok to give it another go and I will look at modularising.

marrobi commented on bootstrap/terraform/build_and_publish_docker_images.sh:
```diff
@@ -0,0 +1,53 @@
+
+echo -e "\e[34m»»» 🔨 \e[96mAzure details from logged on user \e[0m"
+echo -e "\e[34m»»»   • \e[96mSubscription: \e[33m$SUB_NAME\e[0m"
+echo -e "\e[34m»»»   • \e[96mTenant:       \e[33m$TENANT_ID\e[0m\n"
```
Hmm, they are different, this one requires docker...

TessFerrandez replied:
Perhaps we can re-factor out the common code into one script in the interest of keeping things DRY

marrobi replied:
@TessFerrandez  I've just converted to draft, will refactor.

TessFerrandez reviewed (approved):
Approving with suggestion to modularize the deployment scripts

marrobi commented:
@deniscep  @TessFerrandez  I've refactored the scripts, and added a makefile. If follow the doc, then can do everything by running `make xxxx`.

 I need to give it another run through, but let me know your thoughts.

TessFerrandez commented on Makefile:
```diff
@@ -0,0 +1,49 @@
+
+bootstrap:
+	echo -e "\n\e[34m»»» 🧩 \e[96mBootstrap Terraform\e[0m..." \
+	&& . ./devops/scripts/check_dependancies.sh \
```
nit: check_dependencies.sh

TessFerrandez reviewed (approved):
Looks great, just a small typo nit

marrobi commented:
Tested end to end - if check out the repo, edit the env vars in './devops/terraform`, and run `make`, all stages needed to deploy from scratch will run.

christoferlof reviewed (changes requested):
Overall smooth experience 👍, however I'm seeing an Terraform error on first deployment of a new instance.
```
╷
│ Error: Error: Public IP "pip-agw-<name>-<env>-<id>" (Resource Group "rg-<name>-<env>-<id>") was not found
│ 
│   with module.appgateway.data.azurerm_public_ip.appgwpip_data,
│   on appgateway/appgateway.tf line 82, in data "azurerm_public_ip" "appgwpip_data":
│   82: data "azurerm_public_ip" "appgwpip_data" {
│ 
╵
```
I created issue #94 to track and resolve separately.

marrobi commented on docs/quickstart.md:
```diff
@@ -45,29 +32,64 @@ az account list
+- `TF_VAR_state_storage` - The name of the storage account to hold Terraform state.
+- `TF_VAR_mgmt_res_group` - The shared resource group for all hub resources, including the storage account.
+- `TF_VAR_state_container` - Name of the blob container to hold Terraform state (default: `tfstate`).
+- `TF_VAR_prefix` - A prefix added to all resources, pick your project name or other prefix to give the resources unique names.
```
Such as the TRE project name, "hospitalx" or "christofer" for example. Can you thik of better wording?

marrobi commented on docs/quickstart.md:
```diff
@@ -1,40 +1,27 @@
 ![Clone Options](../docs/assets/clone_options.png)
 
-### Pre-requisites
+## Pre-requisites
```
Done
